org.osgi.test.junit4/src/main/java/org/osgi/test/common/annotation/Property.java

```diff
@@ -63,4 +63,6 @@ public enum Type {
 
 	Type type() default Type.Scalar;
 
+	String systemProperty() default NOT_SET;
```

Comment:
What if `value` and `systemProperty` are both set? The current code prefers systemProperty over value but this is not documented. 

Perhaps a better choice is to define the source of the actual value.

```suggestion
    public enum Source { String, SystemProperty, EnvironmentVariable }:
	Source source() default String;
```

`String` is the current behavior. `SystemProperty` means the value holds a system property name. `EnvironmentVariable` means the value holds an env var name.

This would allow for much more future proofing and avoid ambiguity.

Reply:
@bjhargrave if you look here at the original request:
- https://github.com/osgi/osgi-test/issues/614

this was explicitly mean that way that you specify a value that the can be overridden by a system property so this is actually intentional.

If one wants to go your suggestion, then we need some kind of evaluation here instead [like in maven](https://maven.apache.org/pom.html#Properties) that even allows default values (but no `Source`) or like [felix file install](https://felix.apache.org/documentation/subprojects/apache-felix-file-install.html#_property_substitution_in_configuration_files).

I'm just not sure this is a bit over engineered then, the idea was really that plain simple to allow to override the value if a system property of the given name exits, e.g. to specify a different default port for a webserver.

Reply:
I think the change to use a `source` is a good idea, it allows us to support @stbischof in using test properties, and environment variables, without having to add an extra member each time.

The "default" is something that we've lost, but I think it can be easily (and elegantly) supported with a small change to make the second entry in the `value` array be the default if the system property doesn't exist. This would also apply to `EnvironmentVariable`

```
// Defaults to 8080 if web.port.prop is not set
@Property(key="foo", value= {"web.port.prop", "8080"}, source = SystemProperty, scalar = Integer)

// Fails if web.port.prop is not set
@Property(key="foo", value= "web.port.prop", source = SystemProperty, scalar = Integer)

```

Reply:
I just wanted to note that because this PR is marked to fix 
- https://github.com/osgi/osgi-test/issues/614

but nothing of that was requested nor being intended.

I'm not sure what @stbischof request is exactly about but maybe better solved / described in a different topic, I also don't really see how adding a new `SOURCE` type each time is better (what always is exclusive so one cant e.g. use env first, then system, then default) than an attribute ...

Also please note that for env variables there is already a notation of defaults, not sure if felix has something similar:
https://emmer.dev/blog/bash-environment-variable-defaults/

Reply:
@laeubi did you try the felix cm Plugin mechanism

https://github.com/apache/felix-dev/blob/master/configadmin-plugins%2Finterpolation%2FREADME.md

Reply:
I like the way you suggest tim.

// Defaults to 8080 if web.port.prop is not set
@Property(key="foo", value= {"web.port.prop", "8080"}, source = SystemProperty, scalar = Integer)

// Fails if web.port.prop is not set
@Property(key="foo", value= "web.port.prop", source = SystemProperty, scalar = Integer)

Reply:
> I also don't really see how adding a new SOURCE type each time is better

The alternative would have been different members in the `Property` annotation to say `systemProperty` `environmentVariable`, `testClass`, `testName`...
This would explode the apparent complexity of the annotation, and lead to a lot of invalid setups (e.g. setting a `systemProperty` and a `testClass` at the same time).

Adding a `source` gives us one place where we define the rules for getting the value. I am in favour of avoiding extra plugins being needed for simple cases, but would draw the line at trying to replicate bash variable expansion.

Reply:
Well its your to decide, but I don't see any need for testClass (why should it be for?) or env variables, but if it is desired, as said some kind of property expansion would be much more flexible, maybe reusing what @stbischof linked here:

https://github.com/apache/felix-dev/blob/master/configadmin-plugins%2Finterpolation%2FREADME.md

my request only included one simple thing, to override a value, it don't ask for env, testclass testname or any other things, If I would need such high level of felxibility and configuration I would simply not sue an annotation for that...